compiler/rustc_ty_utils/src/ty.rs

```diff
@@ -218,9 +218,12 @@ fn param_env(tcx: TyCtxt<'_>, def_id: DefId) -> ty::ParamEnv<'_> {
     let body_id =
         local_did.and_then(|id| tcx.hir().maybe_body_owned_by(id).map(|body| body.hir_id));
     let body_id = match body_id {
-        Some(id) => id,
-        None if hir_id.is_some() => hir_id.unwrap(),
-        _ => hir::CRATE_HIR_ID,
+        Some(id) => tcx.hir().local_def_id(id),
+        None if hir_id.is_some() => {
+            let id = hir_id.unwrap();
+            tcx.hir().local_def_id(id)
+        }
+        _ => hir::def_id::CRATE_DEF_ID,
     };
```

Comment:
This match changes behaviour.
`local_def_id` will ICE when called on a body id. `local_def_id(body_owned_by(id).hir_id)` is not identity, it's an ICE.
`let body_id = local_id.unwrap_or(CRATE_DEF_ID)` should be enough.

Reply:
Yeah this was the part of the code where I was not sure

Reply:
>let body_id = local_id.unwrap_or(CRATE_DEF_ID) should be enough.

What is `local_id` in this case?

Reply:
The ICE comes from here. I meant `local_did`.

Reply:
Sorry, my bad, I push the fix and it works fine thanks or the catch.

Now I found new ICE and try to fix it in the following comment https://github.com/rust-lang/rust/pull/103902/commits/d6c50c0ea26be8dc2f0c90ea58d0a9adb05a77b6 but I think to exit a better solution for this issues?